src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs

```diff
@@ -211,22 +203,29 @@ internal static unsafe void WriteGatherAtOffset(SafeFileHandle handle, IReadOnly
                     // The write completed successfully but for fewer bytes than requested.
                     // We need to perform next write where the previous one has finished.
                     fileOffset += bytesWritten;
+                    totalBytesToWrite -= bytesWritten;
                     // We need to try again for the remainder.
-                    for (int i = 0; i < buffersCount; i++)
+                    while (buffersOffset < buffersCount && bytesWritten > 0)
                     {
-                        int n = buffers[i].Length;
+                        int n = buffers[buffersOffset].Length;
```

Comment:
Is there anything that could go horribly wrong if the caller-provided IReadOnlyList changed between accesses?

Reply:
this was great question! since we perform a "copy" for the purpose of pinning and IOV, I was able to read that value from the copy.